sol/main.c

```diff
@@ -0,0 +1,568 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <math.h>
+
+#include <rte_sched.h>
+
+#include "gatekeeper_gk.h"
+#include "gatekeeper_launch.h"
+#include "gatekeeper_sol.h"
+
+/*
+ * Gatekeeper request priority queue implementation.
+ *
+ * To implement the request priority queue, we maintain a linked list of
+ * packets listed in order of highest priority to lowest priority. We keep
+ * an array where each index represents a priority, and each element of
+ * that array holds a reference to the last packet of that priority.
+ * This allows us to quickly insert new packets of any priority into
+ * the linked list and drop the packet of lowest priority if necessary.
+ */
+
+static void
+insert_new_priority_req(struct req_queue *req_queue, struct priority_req *pr,
+	uint8_t priority)
+{
+	uint8_t next;
+
+	/* This should be the first request of @priority. */
+	RTE_VERIFY(req_queue->priorities[priority] == NULL);
+
+	req_queue->priorities[priority] = pr;
+
+	/* This is the first packet in the queue. */
+	if (req_queue->len == 0) {
+		list_add(&pr->list, &req_queue->head);
+		req_queue->highest_priority = priority;
+		req_queue->lowest_priority = priority;
+		return;
+	}
+
+	/* Not the first packet, but still insert at the head of the queue. */
+	if (priority > req_queue->highest_priority) {
+		list_add(&pr->list, &req_queue->head);
+		req_queue->highest_priority = priority;
+		return;
+	}
+
+	/*
+	 * Insert in middle or end of queue.
+	 */
+
+	if (priority < req_queue->lowest_priority)
+		req_queue->lowest_priority = priority;
+
+	/*
+	 * This function only inserts a request for a priority that
+	 * does not currently exist, and there is at least one
+	 * request of a higher priority already in the queue.
+	 */
+	RTE_VERIFY(priority != req_queue->highest_priority);
+	for (next = priority + 1; next <= req_queue->highest_priority; next++) {
+		if (req_queue->priorities[next] != NULL) {
+			list_add(&pr->list, &req_queue->priorities[next]->list);
+			return;
+		}
+	}
+
+	rte_panic("sol: %s failed to insert a request of a new priority\n",
+		__func__);
+}
+
+static void
+drop_lowest_priority_pkt(struct sol_config *sol_conf,
+	struct req_queue *req_queue)
+{
+	struct priority_req *lowest_pr = list_last_entry(&req_queue->head,
+		struct priority_req, list);
+	struct priority_req *next_lowest_pr;
+
+	RTE_VERIFY(req_queue->len > 0);
+
+	if (unlikely(req_queue->len == 1)) {
+		req_queue->priorities[lowest_pr->priority] = NULL;
+		req_queue->highest_priority = 0;
+		req_queue->lowest_priority = GK_MAX_REQ_PRIORITY + 1;
```

Comment:
I understand that this and similar cases don't need the plus one, otherwise we shouldn't have zero in `highest_priority`.